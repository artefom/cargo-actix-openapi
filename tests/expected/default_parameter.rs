//! API auto-generated by apigen

use std::fmt::Display;

use std::{collections::HashMap, fmt::Debug};

use serde::{Deserialize, Serialize};

use actix_web::{
    get, http::StatusCode, middleware::NormalizePath, web, App, HttpRequest, HttpResponse,
    HttpServer, ResponseError,
};

use async_trait::async_trait;

// Defaults
// -------------------------------

fn default_str_world() -> String {
    "World".to_string()
}

fn default_float_1() -> f64 {
    1
}

fn default_int_1() -> i64 {
    1
}

fn opt_default_float_1() -> Option<f64> {
    Some(1)
}

fn opt_default_int_1() -> Option<i64> {
    Some(1)
}


// Enums
// -------------------------------


// Struct
// -------------------------------

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct GreetUserPath {
    #[serde(default="default_str_world")]
    pub user: String,
    #[serde(rename="v1_float", default="default_float_1")]
    pub v_1_float: f64,
    #[serde(rename="v1_int", default="default_int_1")]
    pub v_1_int: i64,
    #[serde(rename="v1_opt_float", default="opt_default_float_1")]
    pub v_1_opt_float: Option<f64>,
    #[serde(rename="v1_opt_int", default="opt_default_int_1")]
    pub v_1_opt_int: Option<i64>,
    #[serde(rename="n1", default="opt_default_int_1")]
    pub n_1: Option<i64>,
    #[serde(rename="n2")]
    pub n_2: i64,
    #[serde(rename="n3")]
    pub n_3: Option<i64>,
    #[serde(rename="n4", default="default_int_1")]
    pub n_4: i64,
}


// Error with details
// -------------------------------

/// Bails with detailed api error
#[macro_export]
macro_rules! apibail {
    ($err:expr,$msg:expr) => {
        return Err($crate::server::api::Detailed {
            error: $err,
            details: $msg.to_string(),
        })
    };
}

pub trait StatusCoded {
    fn status_code(&self) -> StatusCode;
}

#[derive(Debug)]
pub struct Detailed<E> {
    pub error: E,
    pub details: String,
}

impl<E: Display> Display for Detailed<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}. Reason: {}", self.error, self.details)
    }
}

impl<E: Display + Debug> std::error::Error for Detailed<E> {}

impl<E: Display + Debug> ResponseError for Detailed<E>
where
    E: StatusCoded,
{
    fn status_code(&self) -> StatusCode {
        self.error.status_code()
    }
}

/// Converts some result to Result with detailed api error
pub trait ApiErr<T, E> {
    /// Wrap the error value with additional context.
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static;
}

impl<T, E> ApiErr<T, E> for Result<T, E>
where
    E: Debug + Send + Sync + 'static,
{
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static,
    {
        // Not using map_err to save 2 useless frames off the captured backtrace
        // in ext_context.
        match self {
            Ok(ok) => Ok(ok),
            Err(original_error) => Err(Detailed {
                error: err,
                details: format!("{:?}", original_error),
            }),
        }
    }
}


// Error
// -------------------------------
