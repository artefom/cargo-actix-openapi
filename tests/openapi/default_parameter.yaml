openapi: "3.0.3"
info:
  version: "1.0"
  title: "Hello World API"
paths:
  /hello/{user}:
    parameters:
      - name: user
        in: path
        required: false
        description: The name of the user to greet.
        schema:
          type: string
          default: "World"
      - name: v1_float
        in: path
        required: false
        schema:
          type: number
          default: 1
      - name: v1_int
        in: path
        required: false
        schema:
          type: integer
          default: 1
      - name: v1_opt_float
        in: path
        required: false
        schema:
          type: number
          default: 1
          nullable: true
      - name: v1_opt_int
        in: path
        required: false
        schema:
          type: integer
          default: 1
          nullable: true
      - name: n1
        in: path
        required: false
        schema:
          type: integer
          default: 1
          nullable: true
      - name: n2
        in: path
        required: true
        schema:
          type: integer
          default: null
          nullable: false
      - name: n3
        in: path
        required: false
        schema:
          type: integer
          default: null
          nullable: true
      - name: n4
        in: path
        required: false
        schema:
          type: integer
          default: 1
          nullable: false
    get:
      operationId: greet_user
      summary: Returns a greeting to the user!
      description: Some detailed description
      responses:
        200:
          description: Returns the greeting.
          content:
            application/json:
              schema:
                type: string
