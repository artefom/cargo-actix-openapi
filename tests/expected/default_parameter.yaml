api:
  definitions:
  - name: default_str_world
    data: !DefaultProvider
      vtype: String
      value: '"World".to_string()'
  - name: default_float_1
    data: !DefaultProvider
      vtype: f64
      value: '1'
  - name: default_int_1
    data: !DefaultProvider
      vtype: i64
      value: '1'
  - name: opt_default_float_1
    data: !DefaultProvider
      vtype: Option<f64>
      value: Some(1)
  - name: opt_default_int_1
    data: !DefaultProvider
      vtype: Option<i64>
      value: Some(1)
  - name: GreetUserPath
    data: !Struct
      doc: null
      properties:
      - name: user
        rename: user
        default: default_str_world
        type_: String
        doc: The name of the user to greet.
      - name: v_1_float
        rename: v1_float
        default: default_float_1
        type_: f64
        doc: null
      - name: v_1_int
        rename: v1_int
        default: default_int_1
        type_: i64
        doc: null
      - name: v_1_opt_float
        rename: v1_opt_float
        default: opt_default_float_1
        type_: Option<f64>
        doc: null
      - name: v_1_opt_int
        rename: v1_opt_int
        default: opt_default_int_1
        type_: Option<i64>
        doc: null
      - name: n_1
        rename: n1
        default: opt_default_int_1
        type_: Option<i64>
        doc: null
      - name: n_2
        rename: n2
        default: null
        type_: i64
        doc: null
      - name: n_3
        rename: n3
        default: null
        type_: Option<i64>
        doc: null
      - name: n_4
        rename: n4
        default: default_int_1
        type_: i64
        doc: null
  operations:
  - name: greet_user
    path: /hello/{user}
    method: Get
    doc: Returns a greeting to the user!
    param_path: web::Path<GreetUserPath>
    param_query: null
    param_body: null
    response: web::Json<String>
